src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt

```diff
@@ -0,0 +1,403 @@
+     *
+     * @param clazz the class to transform
+     */
+    private fun ensureClassHierarchyIsTransformed(clazz: Class<*>) {
+        if (INSTRUMENT_ALL_CLASSES_IN_MODEL_CHECKING_MODE) {
+            return
+        }
+        if (clazz.name in instrumentedClassesInTheModelCheckingMode) return // already instrumented
+        ensureClassHierarchyIsTransformed(clazz, Collections.newSetFromMap(IdentityHashMap()))
+    }
+
+    /**
+     * Ensures that the given object and all its referenced objects are transformed according to the provided rules.
+     * The transformation is performed recursively, starting from the given object.
+     *
+     * @param obj The object to be ensured for transformation.
+     * @param processedObjects A set of processed objects to avoid infinite recursion.
+     */
+    private fun ensureObjectIsTransformed(obj: Any, processedObjects: MutableSet<Any>) {
+        if (!instrumentation.isModifiableClass(obj.javaClass) || !shouldTransform(obj.javaClass.name, instrumentationMode)) {
+            return
+        }
+
+        if (processedObjects.contains(obj)) return
+        processedObjects += obj
+
+        var clazz: Class<*> = obj.javaClass
+
+        ensureClassHierarchyIsTransformed(clazz)
+
+        while (true) {
+            clazz.declaredFields
+                .filter { !it.type.isPrimitive }
+                .filter { !Modifier.isStatic(it.modifiers) }
+                .mapNotNull { readFieldViaUnsafe(obj, it) }
+                .forEach {
+                    ensureObjectIsTransformed(it, processedObjects)
+                }
+            clazz = clazz.superclass ?: break
+        }
+    }
+
+    /**
+     * Ensures that the given class and all its superclasses are transformed.
+     *
+     * @param clazz The class to be transformed.
+     * @param processedObjects Set of objects that have already been processed to prevent duplicate transformation.
+     */
+    private fun ensureClassHierarchyIsTransformed(clazz: Class<*>, processedObjects: MutableSet<Any>) {
+        if (instrumentation.isModifiableClass(clazz) && shouldTransform(clazz.name, instrumentationMode)) {
+            instrumentedClassesInTheModelCheckingMode += clazz.name
+            instrumentation.retransformClasses(clazz)
+        } else {
+            return
+        }
+        // Traverse static fields.
+        clazz.declaredFields
+            .filter { !it.type.isPrimitive }
+            .filter { Modifier.isStatic(it.modifiers) }
+            .mapNotNull { readFieldViaUnsafe(null, it) }
+            .forEach {
+                ensureObjectIsTransformed(it, processedObjects)
+            }
+        clazz.superclass?.let {
+            if (it.name in instrumentedClassesInTheModelCheckingMode) return // already instrumented
+            ensureClassHierarchyIsTransformed(it, processedObjects)
+        }
+    }
+
+    private fun readFieldViaUnsafe(obj: Any?, field: Field): Any? =
```

Comment:
The same code is now used for the plugin, please keep only one version.

Reply:
Removed duplicate implementation, and put `readFieldViaUnsafe` into `util` package.